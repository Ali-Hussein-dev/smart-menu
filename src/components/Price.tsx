import { ImInfo } from 'react-icons/im'
interface Types {
  price: number
  visible: boolean
  setVisible: (visible: boolean) => void
}
//=======================
const Price: React.FC<Types> = ({ price, setVisible, visible }) => {
  // hooks
  //--------------------------------------
  // functions
  //--------------------------------------
  return (
    <div className="flex flex-col items-center">
      {' '}
      <span>{price.toFixed(2)}</span>
      <button
        aria-label={`info-${price}`}
        id={`info-${price}`}
        onClick={() => {
          setVisible(!visible)
        }}
        role="button"
        className="p-1"
      >
        <ImInfo size="15" />
      </button>
    </div>
  )
}
export default Price
