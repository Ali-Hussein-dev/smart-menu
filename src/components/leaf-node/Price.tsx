import { ImInfo } from 'react-icons/im'
import * as React from 'react'
import { Ctx } from '../../context'
import { Divider } from '@chakra-ui/react'

const ToFixed: React.FC<{
  num: number | undefined
  unit?: Menu.SizeUnit
  size?: number
  isHH?: boolean
  currency?: "€" | "$"
}> = ({ num, unit, size, isHH = false, currency = '€' }) => (
  <span className="flex justify-center gap-x-1">
    {typeof num === 'number' && isHH ? (
      <span className="text-sm line-through text-blueGray-400">
        {num.toFixed(2)}
      </span>
    ) : (
      <span>
        {typeof num === 'number' && num.toFixed(2)}
        <span className="pl-1 text-blueGray-500">{currency}</span>
      </span>
    )}
    {(size || unit) && (
      <span className="text-blueGray-400">
        {size}
        {size ? unit : null}
      </span>
    )}
  </span>
)
interface PriceProps {
  price: Menu.Price
  size?: Menu.Size
  isOpen?: boolean
  isInfo?: boolean
  hasHH?: boolean
  sizeUnit?: Menu.SizeUnit
  onToggle?: (isOpen: boolean) => void
}
//=======================
export const Price: React.FC<PriceProps> = ({
  price,
  size,
  onToggle,
  isOpen,
  isInfo = false,
  sizeUnit,
  hasHH,
}) => {
  // hooks
  const { state } = React.useContext(Ctx)
  const { meta } = state
  //--------------------------------------
  // functions
  //--------------------------------------
  return (
    <div className="flex flex-col items-center justify-between ml-1 text-blueGray-600">
      <div className="flex items-center justify-center">
        {hasHH && meta.isHH && (
          <span className="flex flex-col mr-1">
            {price.pp1 && <ToFixed num={price.pp1} />}
            {price.pp2 && <ToFixed num={price.pp2} />}
          </span>
        )}

        {hasHH ? ( // ----hh price
          <span className="flex flex-col">
            {price.p1 && (
              <ToFixed
                isHH={meta.isHH}
                num={price.p1}
                size={size?.s1}
                unit={sizeUnit}
              />
            )}
            {price.p2 && (
              <ToFixed isHH={meta.isHH} num={price.p2} unit={sizeUnit} />
            )}
          </span>
        ) : (
          <span className="flex flex-col">
            {price.p1 && (
              <ToFixed num={price.p1} size={size?.s1} unit={sizeUnit} />
            )}
            {price.p2 && (
              <ToFixed num={price.p2} size={size?.s2} unit={sizeUnit} />
            )}
          </span>
        )}
      </div>
      <div className="flex flex-col items-center w-full">
        {isInfo && (
          <button
            aria-label="item-info"
            onClick={() => {
              onToggle(!isOpen)
            }}
            role="button"
            className="pb-1"
          >
            <ImInfo size="15" className="text-blue-300 " />
          </button>
        )}
        <Divider variant="dashed" />
      </div>
    </div>
  )
}
