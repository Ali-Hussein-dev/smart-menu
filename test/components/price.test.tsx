import userEvent from '@testing-library/user-event'
import { screen, render } from '@testing-library/react'
import Price from '../../src/components/Price'
describe('Price', () => {
  // eslint-disable-next-line prefer-const
  let visible = false
  const setVisible = () => {
    visible = !visible
  }
  test('setVisible should be called', () => {
    const mockSetVisible = jest.fn(setVisible)
    const { container, asFragment, debug } = render(
      <Price price={3} visible={visible} setVisible={mockSetVisible} />
    )
    asFragment()
    debug(container)
    userEvent.click(screen.getByLabelText('info-3'))
    expect(mockSetVisible).toBeCalled()
    expect(mockSetVisible).toBeCalledWith(true)
  })

  test('price matches textContent', () => {
    const { container } = render(
      <Price price={3} visible={visible} setVisible={() => setVisible()} />
    )
    expect(container.textContent).toMatch('3.00')
  })
  test('price snapshot', () => {
    const { container } = render(
      <Price price={3} visible={visible} setVisible={() => setVisible()} />
    )
    expect(container.textContent).toMatchSnapshot()
  })
})
