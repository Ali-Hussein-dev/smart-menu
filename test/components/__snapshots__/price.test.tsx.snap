// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price Is info button visible 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-between ml-1 text-blueGray-600"
  >
    <div
      class="flex items-center justify-center"
    >
      <span
        class="flex flex-col"
      >
        <span
          class="flex justify-center gap-x-1"
        >
          <span>
            9.50
            <span
              class="pl-1 text-blueGray-500"
            >
              €
            </span>
          </span>
        </span>
        <span
          class="flex justify-center gap-x-1"
        >
          <span>
            12.50
            <span
              class="pl-1 text-blueGray-500"
            >
              €
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      class="flex flex-col items-center w-full"
    >
      <button
        aria-label="item-info"
        class="pb-1"
        role="button"
      >
        <svg
          class="text-blue-300 "
          fill="currentColor"
          height="15"
          stroke="currentColor"
          stroke-width="0"
          version="1.1"
          viewBox="0 0 16 16"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 4.75c0-0.412 0.338-0.75 0.75-0.75h0.5c0.412 0 0.75 0.338 0.75 0.75v0.5c0 0.412-0.338 0.75-0.75 0.75h-0.5c-0.412 0-0.75-0.338-0.75-0.75v-0.5z"
          />
          <path
            d="M10 12h-4v-1h1v-3h-1v-1h3v4h1z"
          />
          <path
            d="M8 0c-4.418 0-8 3.582-8 8s3.582 8 8 8 8-3.582 8-8-3.582-8-8-8zM8 14.5c-3.59 0-6.5-2.91-6.5-6.5s2.91-6.5 6.5-6.5 6.5 2.91 6.5 6.5-2.91 6.5-6.5 6.5z"
          />
        </svg>
      </button>
      <hr
        aria-orientation="horizontal"
        class="chakra-divider css-1647zlw"
        role="separator"
      />
    </div>
  </div>
</div>
`;

exports[`Price Is price visible 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-between ml-1 text-blueGray-600"
  >
    <div
      class="flex items-center justify-center"
    >
      <span
        class="flex flex-col"
      >
        <span
          class="flex justify-center gap-x-1"
        >
          <span>
            9.50
            <span
              class="pl-1 text-blueGray-500"
            >
              €
            </span>
          </span>
        </span>
        <span
          class="flex justify-center gap-x-1"
        >
          <span>
            12.50
            <span
              class="pl-1 text-blueGray-500"
            >
              €
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      class="flex flex-col items-center w-full"
    >
      <hr
        aria-orientation="horizontal"
        class="chakra-divider css-1647zlw"
        role="separator"
      />
    </div>
  </div>
</div>
`;

exports[`Price Is price visible with size & units 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-between ml-1 text-blueGray-600"
  >
    <div
      class="flex items-center justify-center"
    >
      <span
        class="flex flex-col"
      >
        <span
          class="flex justify-center gap-x-1"
        >
          <span>
            9.50
            <span
              class="pl-1 text-blueGray-500"
            >
              €
            </span>
          </span>
          <span
            class="text-blueGray-400"
          >
            0.3
            L
          </span>
        </span>
        <span
          class="flex justify-center gap-x-1"
        >
          <span>
            12.50
            <span
              class="pl-1 text-blueGray-500"
            >
              €
            </span>
          </span>
          <span
            class="text-blueGray-400"
          >
            0.5
            L
          </span>
        </span>
      </span>
    </div>
    <div
      class="flex flex-col items-center w-full"
    >
      <hr
        aria-orientation="horizontal"
        class="chakra-divider css-1647zlw"
        role="separator"
      />
    </div>
  </div>
</div>
`;
